allennlp/service/config_explorer.html

```diff
@@ -0,0 +1,608 @@
+            {renderedAnnotation}
+            {renderedDefault}
+            {renderChild(path, item, setData)}
+        </div>
+    )
+}
+
+// Find the longest common prefix
+const sharedPrefix = (s1, s2) => {
+    for (let i = 0; i < s1.length; i++) {
+        if (s1[i] !== s2[i]) {
+            return s1.slice(0, i)
+        }
+    }
+    // Otherwise, all of s1
+    return s1
+}
+
+const commonPrefix = (strings) => {
+    if (strings.size == 1) {
+        // Special case, return up to last .
+        const [string] = strings
+        const idx = string.lastIndexOf(".")
+        return string.slice(0, idx + 1)
+    } else {
+        return strings.reduce(sharedPrefix)
+    }
+}
+
+const Configurator = ({path, item, setData}) => {
+    const annotation = item.get('annotation')
+    const className = annotation.get('origin')
+
+    // User-supplied
+    const choices = item.get('choices')
+    const choice = item.get('choice')
+    const config = item.get('value')
+
+    const prefix = choices ? commonPrefix(choices) : null
+
+    const getChoices = () => {
+        fetch('/api/?class=' + className)
+            .then(res => res.json())
+            .then(({config, choices}) => {
+                if (choices) {
+                    setData(oldConfig => oldConfig.setIn(path.push('choices'), Immutable.List(choices)))
+                } else {
+                    setData(oldConfig => oldConfig.setIn(path.push('value'), Immutable.fromJS(config)))
+                }
+            })
+    }
+
+    const select = (evt) => {
+        const choice = evt.target.value
+
+        fetch('/api/?class=' + choice)
+            .then(res => res.json())
+            .then(({config}) => {
+                setData(rootConfig => rootConfig.setIn(path.push('choice'), choice)
+                                                .setIn(path.push('value'), Immutable.fromJS(config)))
+            })
+    }
+
+    const remove = () => {
+        setData(config => config.deleteIn(path.push('choices'))
+                                .deleteIn(path.push('value'))
+                                .deleteIn(path.push('choice')))
+    }
+
+    const configureButton = (choices || config) ? null : <button onClick={getChoices}>CONFIGURE</button>
+    const choicesDropdown = choices ? (
+        <span>
+            <span class="prefix">{prefix}</span>
+            <select value={choice} onChange={select}>
+                {choice ? null : (<option value=""></option>)}
+                {choices.map(subclass => <option value={subclass}>{subclass.slice(prefix.length)}</option>)}
+            </select>
+        </span>
+    ) : null
+    const deleteButton = choices ? <button onClick={remove}>x</button> : null
+    const renderedValue = config ? <Config path={path.push('value')} config={config} setData={setData}/> : null
+
+    return <span>
+            {configureButton}
+            <div class="choices-dropdown">
+                {choicesDropdown}{deleteButton}{renderedValue}
```

Comment:
Nit: render the `deleteButton` first so that it's always in the same place - currently, for `List` items, it is before the entry field, but here it's after.

Reply:
I don't like this way, but if I make that change then a `Dict[str, Configurable]` element becomes a huge mess